pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,405 @@
+func (k *Kad) recalcDepth() uint8 {
+	// handle edge case separately
+	if k.connectedPeers.Length() <= nnLowWatermark {
+		return 0
+	}
+	var (
+		peers                        = uint(0)
+		candidate                    = uint8(0)
+		shallowestEmpty, noEmptyBins = k.connectedPeers.ShallowestEmpty()
+	)
+
+	_ = k.connectedPeers.EachBin(func(_ swarm.Address, po uint8) (bool, bool, error) {
+		peers++
+		if peers >= nnLowWatermark {
+			candidate = po
+			return true, false, nil
+		}
+		return false, false, nil
+	})
+
+	if noEmptyBins || shallowestEmpty > candidate {
+		return candidate
+	}
+
+	return shallowestEmpty
+}
+
+// AddPeer adds a peer to the knownPeers list.
+// This does not guarantee that a connection will immediately
+// be made to the peer.
+func (k *Kad) AddPeer(ctx context.Context, addr swarm.Address) error {
+	if k.connectedPeers.Exists(addr) || k.knownPeers.Exists(addr) {
+		return nil
+	}
+
+	po := swarm.Proximity(k.base.Bytes(), addr.Bytes())
+	k.knownPeers.Add(addr, uint8(po))
+
+	select {
+	case k.manageC <- struct{}{}:
+	default:
+	}
+
+	return nil
+}
+
+// Disconnected is called when peer disconnects.
+func (k *Kad) Disconnected(addr swarm.Address) {
+	po := uint8(swarm.Proximity(k.base.Bytes(), addr.Bytes()))
+	k.connectedPeers.Remove(addr, po)
+
+	k.waitNextMu.Lock()
+	k.waitNext[addr.String()] = time.Now().Add(timeToRetry)
+	k.waitNextMu.Unlock()
+
+	k.depthMu.Lock()
+	k.depth = k.recalcDepth()
+	k.depthMu.Unlock()
+	select {
+	case k.manageC <- struct{}{}:
+	default:
+	}
+}
+
+// ClosestPeer returns the closest peer to a given address.
+func (k *Kad) ClosestPeer(addr swarm.Address) (peerAddr swarm.Address, err error) {
+	panic("not implemented") // TODO: Implement
+}
+
+// NeighborhoodDepth returns the current Kademlia depth.
+func (k *Kad) NeighborhoodDepth() uint8 {
+	k.depthMu.RLock()
+	defer k.depthMu.RUnlock()
+
+	return k.neighborhoodDepth()
+}
+
+func (k *Kad) neighborhoodDepth() uint8 {
+	return k.depth
+}
+
+// MarshalJSON returns a JSON representation of Kademlia.
+func (k *Kad) MarshalJSON() ([]byte, error) {
+	return k.marshal(false)
+}
+
+func (k *Kad) marshal(indent bool) ([]byte, error) {
+	type binInfo struct {
+		BinPopulation  uint
+		BinConnected   uint
+		KnownPeers     []string
+		ConnectedPeers []string
+	}
+
+	type kadBins struct {
+		Bin0  binInfo `json:"bin_0"`
+		Bin1  binInfo `json:"bin_1"`
+		Bin2  binInfo `json:"bin_2"`
+		Bin3  binInfo `json:"bin_3"`
+		Bin4  binInfo `json:"bin_4"`
+		Bin5  binInfo `json:"bin_5"`
+		Bin6  binInfo `json:"bin_6"`
+		Bin7  binInfo `json:"bin_7"`
+		Bin8  binInfo `json:"bin_8"`
+		Bin9  binInfo `json:"bin_9"`
+		Bin10 binInfo `json:"bin_10"`
+		Bin11 binInfo `json:"bin_11"`
+		Bin12 binInfo `json:"bin_12"`
+		Bin13 binInfo `json:"bin_13"`
+		Bin14 binInfo `json:"bin_14"`
+		Bin15 binInfo `json:"bin_15"`
+	}
+
+	type kadParams struct {
+		Base           string    // base address string
+		Population     int       // known
```

Comment:
briefly:
population - all peers we know about in a bin (disconnected + connected)
connected - number of peers we are currently connected to

so `known` because `all we know of`

Reply:
thanks for the clarification